hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/StoreSnapshotTasklet.java

```diff
@@ -165,4 +183,22 @@ public String toString() {
         /** Input is done, terminal state. */
         DONE
     }
+
+    private static class Metrics implements DynamicMetricsProvider {
+
+        private long bytes;
+        private long keys;
+
+        synchronized void set(long bytes, long keys) {
+            this.bytes = bytes;
+            this.keys = keys;
+        }
+
+        @Override
+        public synchronized void provideDynamicMetrics(MetricDescriptor descriptor, MetricsCollectionContext context) {
+            context.collect(descriptor, SNAPSHOT_BYTES, ProbeLevel.INFO, ProbeUnit.COUNT, bytes);
+            context.collect(descriptor, SNAPSHOT_KEYS, ProbeLevel.INFO, ProbeUnit.COUNT, keys);
```

Comment:
Why use `synchronized` and not just `volatile`? To avoid observing keys updated, but not bytes? This problem is all over the metrics, I wouldn't be bothered by that.

Reply:
Mostly because they're updated together and in very low frequency (once every few sec), I don't know if there's something else that's comparable? There's still a very slim chance you get mismatching values, it might look confusing and won't be updated until next snapshot.

Reply:
We never synchronize any metrics. For example, you can observe `totalKeys` zero and `totalWindows` non-zero. Or you can see 1 received item in a downstream processor and 0 emitted at upstream processor (and the downstream processor didn't generate the item). Even using lazySet can cause that.

I'm not bothered by this at all. We're constantly turning metrics into an application logic thing that must be perfectly reliable and consistent. Metrics give you a peep at the engine's entrails. Compare it to car's diagnostics, each value you see is a bit late and each is a different bit late.

Also the all snapshots tend to be similarly large, so seeing two values each from a different snapshot will not be immediately apparent.